cloud.foundry.cli/src/main/java/cloud/foundry/cli/operations/ServicesOperations.java

```diff
@@ -120,8 +129,162 @@ public void updateServiceInstance(String serviceInstanceName, ServiceBean servic
+
+            // unbind apps
+            unbindApps(serviceInstanceName, serviceInstance);
+
+            // delete keys
+            deleteKeys(serviceInstanceName, serviceInstance);
+
+            // do delete service instance
+            DeleteServiceInstanceRequest deleteServiceInstanceRequest = DeleteServiceInstanceRequest
+                .builder()
+                .name(serviceInstanceName)
+                .build();
+
+            this.cloudFoundryOperations.services()
+                .deleteInstance(deleteServiceInstanceRequest)
+                .block();
+
+            Log.info("Service " + serviceInstanceName + " has been removed.");
+      
+        } catch (RuntimeException e) {
+            throw new IllegalArgumentException(e.getMessage());
+        }
+    }
+
+    /**
+     * Delete service keys.
+     *
+     * @param serviceInstanceName serviceInstanceName Name of a service instance.
+     * @param serviceInstance     A service instance.
+     */
+    private void deleteKeys(String serviceInstanceName, ServiceInstance serviceInstance) {
+        if (!serviceInstance.getType().getValue().equals(USER_PROVIDED_SERVICE_INSTANCE)) {
+            ListServiceKeysRequest listServiceKeysRequest = ListServiceKeysRequest
+                .builder()
+                .serviceInstanceName(serviceInstanceName)
+                .build();
+
+            try {
+                this.cloudFoundryOperations
+                    .services()
+                    .listServiceKeys(listServiceKeysRequest)
+                    .flatMap(key -> doDeleteKey(serviceInstanceName, key))
+                    .collectList()
+                    .block();
+                Log.info("All service keys of service instance " + serviceInstanceName + " have been deleted.");
+          
+            } catch (Exception e) {
+                Log.error(e.getMessage());
+            }
+        }
+    }
+
+    private Mono<Void> doDeleteKey(String serviceInstanceName, ServiceKey key) {
+        return this.cloudFoundryOperations
+            .services()
+            .deleteServiceKey(DeleteServiceKeyRequest.builder()
+                .serviceInstanceName(serviceInstanceName)
+                .serviceKeyName(key.getName())
+                .build());
+    }
+
+    /**
+     * Unbind a service instance from all applications.
+     *
+     * @param serviceInstanceName serviceInstanceName Name of a service instance.
+     * @param serviceInstance     A service instance.
+     */
+    private void unbindApps(String serviceInstanceName, ServiceInstance serviceInstance) {
+        if (serviceInstance.getApplications() == null || serviceInstance.getApplications().isEmpty()) {
+            Log.info("There is no application to unbind!");
+        } else {
+            try {
+                serviceInstance.getApplications()
+                    .parallelStream()
+                    .forEach(applicationName -> doUnbindApp(serviceInstanceName, applicationName));
```

Comment:
Might be more consistent to use Reactor here also. Flux.fromIterable(...). But that's ok for now. :)